state-synchronizer/src/tests/fuzzing.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    chunk_request::{GetChunkRequest, TargetType},
+    chunk_response::{GetChunkResponse, ResponseLedgerInfo},
+    coordinator::SyncCoordinator,
+    network::{StateSynchronizerMsg, StateSynchronizerSender},
+    tests::{
+        helpers::{MockExecutorProxy, SynchronizerEnvHelper},
+        mock_storage::MockStorage,
+    },
+};
+use channel::{libra_channel, message_queues::QueueStyle};
+use futures::channel::mpsc;
+use libra_config::{
+    config::{NodeConfig, PeerNetworkId, RoleType},
+    network_id::{NetworkId, NodeNetworkId},
+};
+use libra_proptest_helpers::ValueGenerator;
+use libra_types::{
+    ledger_info::LedgerInfoWithSignatures, transaction::TransactionListWithProof,
+    waypoint::Waypoint, PeerId,
+};
+use network::{
+    peer_manager::{ConnectionRequestSender, PeerManagerRequestSender},
+    protocols::network::NewNetworkSender,
+};
+use proptest::{
+    arbitrary::{any, Arbitrary},
+    option,
+    prelude::*,
+    strategy::Strategy,
+};
+use std::{
+    collections::HashMap,
+    num::NonZeroUsize,
+    sync::{Arc, RwLock},
+};
+
+#[test]
+fn test_state_sync_fuzzer() {
+    let mut gen = ValueGenerator::new();
+    let data = generate_corpus(&mut gen);
+    fuzzer(data);
+}
+
+pub fn generate_corpus(gen: &mut ValueGenerator) -> Vec<u8> {
+    let msg = gen.generate(state_sync_msg_strategy());
+    lcs::to_bytes(&msg).expect("[fuzzing] failed to generate corpus for state sync msg")
+}
+
+pub fn fuzzer(data: Vec<u8>) {
+    let msg = match lcs::from_bytes::<StateSynchronizerMsg>(&data) {
+        Ok(msg) => msg,
+        Err(_) => return,
+    };
+
+    // start up coordinator
+    let (_coordinator_sender, coordinator_receiver) = mpsc::unbounded();
+    let (mempool_sender, _mempool_receiver) = mpsc::channel(1_024);
```

Comment:
so two thoughts:

- maybe using these large buffers or unbounded channels will slow down the fuzzer
- since the receiver is not used, there's the possibility of the fuzzer blocking (I'm guessing that's why the first one is unbounded)

I think a better approach would be to have fake channels where sending is just a no-op. That might be tricky to do though since a sender doesn't implement a sending trait... food for thoughts

Reply:
SyncCoordinator requires a very specific type of channel, so I don't think I can write a mocked channel I can switch it out for without actually changing the SyncCoordinator implementation...

Reply:
yeah I ran into this problem elsewhere. I really don't understand why a `Sender` channel wouldn't implement a `TSender` trait : c

I think I've seen workarounds where people use a certain type depending on the flags. For a channel it would look like this perhaps:

```rust
#[cfg(not(feature = "fuzzing"))]
type Sender = mpsc::Sender;
#[cfg(feature = "fuzzing"))]
type Sender = fakeSender;
```

a bit ugly though :P

Reply:
I'll explore and put out a follow-up PR if I come up with something good.. but re: the receiver not being used, that channel state sync actually sets a timeout on, so it won't hang forever, so the fuzzer at least won't be hanging for that